File: data/src/tfedb/dice.clj
(ns tfedb.dice)

(defn ->dice [number sides]
  {:number number
   :sides sides})

(defn int->dice [i]
  {:number (bit-and i 0x3F)
   :plus (bit-shift-right i 18)
   :sides (-> i
              (bit-shift-right 6)
              (bit-and 0xFFF))})

(defn dice->str [{:keys [number plus sides] :as dice}]
  (when dice
    (cond->
      (str number "d" sides)
      plus
      (str "+" plus))))

(defn dice-range [{:keys [number plus sides]}]
  [(cond-> sides plus (+ plus))
   (cond-> (* number sides) plus (+ plus))])

File: ui/src/format/dice.ts
import * as entity from '../types/entity';

export function prettyDice(dice: entity.Dice) {
  if (!dice) {
    return null;
  }
  if (dice.number === 0 || !dice.sides) {
    return '0';
  }
  let s = `${dice.number}d${dice.sides}`;
  let from = dice.number;
  let to = dice.number * dice.sides;
  if (dice.plus) {
    s = `${s}+${dice.plus}`;
    from += dice.plus;
    to += dice.plus;
  }
  s = `${s} (${from}-${to})`;
  return s;
}
